### ISSUE TITLE
dbt.exceptions.InternalException: Asked to compile <class 'dbt.contracts.graph.compiled.CompiledModelNode'> node, but it has no compiled form

### ISSUE THREAD

**Issue by reubster** [sqlfluff/sqlfluff#3587]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

While linting any dbt model the error is raised part-way through linting (apparently after all the fail lines have been produced and displayed). 

I've tried sqlfluff versions 

0.10.1 - no error 
0.11.0, 0.13.2, 1.0.0, 1.1.0 - error thrown

### Expected Behaviour

sqlfluff reports linting failures. 

### Observed Behaviour

SQL model materialising as a view. This is really the WEB_USERS model from the fivetran segment dbt package, just renamed to use our naming standard. 

```
with

SOURCE as (
  select
    *
  from
    {{ ref('SG__WEB_USERS__UNION') }}
),

RENAMED as (
  select
    _DBT_SOURCE_RELATION,
    EMAIL,

    {{ sv_utils.get_email_domain('EMAIL') }} as EMAIL_DOMAIN,

    {{ sv_utils.is_valid_email('EMAIL') }} as IS_EMAIL_VALID,

    ID as USER_ID,
    PERSON_ID,
    USERNAME,
    EMAIL_ADDRESS,
    {{ sv_utils.get_email_domain('EMAIL_ADDRESS') }} as EMAIL_ADDRESS_DOMAIN,
    {{ sv_utils.is_valid_email('EMAIL_ADDRESS') }} as IS_EMAIL_ADDRESS_VALID,
    "NAME" as FULL_NAME,
    SHORT_NAME,
    FIRST_NAME,
    INITIALS,
    SURNAME,
    LAST_NAME,
    USER_TYPE,

    CONTEXT_IP as IP,
    -- CONTEXT_LIBRARY_NAME, -- not useful
    SUBSCRIPTIONS,
    CONTEXT_PAGE_SEARCH as PAGE_URL_QUERY,
    CONTEXT_PAGE_URL as PAGE_URL,
    {{ dbt_utils.get_url_host('CONTEXT_PAGE_URL') }} as PAGE_URL_HOST,
    -- CONTEXT_LIBRARY_VERSION, -- not useful
    CONTEXT_PAGE_REFERRER as PAGE_REFERRER,

    replace(
      {{ dbt_utils.get_url_host('CONTEXT_PAGE_REFERRER') }},
      'www.',
      ''
    ) as PAGE_REFERRER_HOST,

    CONTEXT_PAGE_TITLE as PAGE_TITLE,
    CONTEXT_LOCALE as LOCALE,
    CONTEXT_PAGE_PATH as PAGE_PATH,
    CONTEXT_CAMPAIGN_SOURCE as UTM_SOURCE,
    CONTEXT_CAMPAIGN_MEDIUM as UTM_MEDIUM,
    CONTEXT_CAMPAIGN_NAME as UTM_CAMPAIGN,
    BISLR_ID,
    CONTEXT_USER_AGENT as USER_AGENT,
    case
      when lower(context_user_agent) like '%android%' then 'Android'
      else replace(
        {{ dbt_utils.split_part(
          dbt_utils.split_part('CONTEXT_USER_AGENT',"'('",2),"' '",1) 
        }},
        ';','')
    end as device,
    UUID_TS,
    RECEIVED_AT as RECEIVED_AT_DATE_TIME,
    "ROLE" as COMPANY_ROLE,
    AGENCY_ID,
    ORGANISATION_ID,
    COMPANY_ID,
    COMPANY,
    AGENCY_NAME,
    ORGANISATION_NAME,
    COMPANY_NAME,
    ORGANISATION_TRADING_STYLES,
    COMPANY_FINANCIAL_FORCE_ID,
    COMPANY_CURRENCIES,
    COMPANY_RISK_PROFILES,

    AGENCIES,
    HAS_LTD_TERMS,
    AGENCY_ACCOUNT_STATUS,
    PROPOSAL_STATE,
    HAS_LABS_ACCESS,
    HAS_PAYE_TERMS,
    AGENCY_PAYROLL_PROVIDER,
    AML_STATUS,
    AGENCY_SIGN_UP_STATE,
    TIME_SHEETS_AVAILABLE,
    PICTURE_URL,
    AGENCY_HAS_FULL_ACCESS,
    ONLINE_TIMESHEET_PROVIDER,

    ROLES,
    ROLES_IS_SUPER_USER,
    ROLES_IS_CLIENT_AUTHORISER,
    ROLES_IS_SONOVATE_ADMIN,
    ROLES_IS_CANDIDATE,
    PERMISSIONS as USERS_PERMISSIONS,
    PERMISSIONS_IS_CLIENT_AUTHORISER,
    PERMISSIONS_IS_SONOVATE_ADMIN,
    PERMISSIONS_IS_CANDIDATE,

    "PROFILE" as USERS_PROFILE,
    "SESSION_ID" as USERS_SESSION_ID,
    AUTHENTICATED,
    IMAGE_URL,
    AUTHENTICATION_TOKEN,
    TOKEN,
    PERSON_TOKEN,

    LANGUAGE_CODE,
    SETTINGS_DATE_FORMAT,
    SETTINGS_SUBSCRIPTIONS,
    SETTINGS_TIME_ZONE,
    SETTINGS_LANGUAGE,
    SETTINGS_TIME_FORMAT,
    SETTINGS_COUNTRY_CODE,
    SETTINGS_WEEK_START

  from
    SOURCE
),

FINAL as (
  select 
    *,
    case
      when DEVICE = 'iPhone' then 'iPhone'
      when DEVICE = 'Android' then 'Android'
      when DEVICE in ('iPad', 'iPod') then 'Tablet'
      when DEVICE in ('Windows', 'Macintosh', 'X11')
        then 'Desktop'
      else 'Uncategorized' end as DEVICE_CATEGORY
  from
    RENAMED
)

select
  *
from
  FINAL

```

Example verbose output 

[fluff.txt](https://github.com/sqlfluff/sqlfluff/files/9083364/fluff.txt)

### How to reproduce

Just a basic sqlfluff lint operation 

sqlfluff lint .\models\base\segment\SG__WEB_USERS.sql 

### Dialect

snowflake

### Version

0.11.0
0.13.2
1.0.1
1.1.0

### Configuration

.sqlfluff

```
# dbt templating does not keep trailing new lines (L009)

[sqlfluff]
verbose = 3
templater = dbt
dialect = snowflake
exclude_rules = L009,L031,L034
ignore_templated_areas = True
sql_file_exts = .sql,.sql.j2,.dml,.ddl,.SQL

[sqlfluff:rules]
tab_space_size = 2

[sqlfluff:rules:L010]
capitalisation_policy = lower

[sqlfluff:rules:L014]
extended_capitalisation_policy = upper

[sqlfluff:rules:L016]
max_line_length = 180

[sqlfluff:rules:L019]
comma_style = leading

[sqlfluff:rules:L030]
extended_capitalisation_policy = lower

[sqlfluff:rules:L040]
capitalisation_policy = lower

[sqlfluff:templater:dbt]
apply_dbt_builtins = True
load_macros_from_path=./dbt_modules/dbt_utils/macros,./macros

[sqlfluff:templater:dbt:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

[sqlfluff:templater:jinja:context]
num_things=456
tbl_name=my_table
my_list=['a', 'b', 'c']
MY_LIST=("d", "e", "f")

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
load_macros_from_path=./dbt_modules/dbt_utils/macros

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

```

.sqlfluffignore

```
target/
dbt_modules/
macros/

```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tsingh2k15** [comment]
I am running into exact same issue, is there a workaround available? When can we expect a fix?

**Comment by barrywhart** [comment]
This is free software developed and maintained by volunteers. Asking when to "expect" a fix is a bit presumptuous. Are you interested in contributing a fix?

**Comment by tsingh2k15** [comment]
I apologize for that. I will investigate it further locally and update this thread with my findings.

**Comment by tsingh2k15** [comment]
One observation:  In my case, the error occurs while `sorting nodes` on my `staging` models. Other models are passing `sorting nodes`. I will continue to gather more details.

This is the details of my environment. 

```
==== sqlfluff ====
sqlfluff:                1.1.0 python:                 3.8.11
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:                 dbt
dbt:                    =1.1.1
rules:                                all
```

**Comment by barrywhart** [comment]
Sorting nodes is a step that determines what order to lint the models. IIRC, in order to lint ephemeral models, it is necessary to first lint at least one other model that depends on the ephemeral model. Although you mentioned there are no ephemeral models in your project, ephemeral models are the only thing I know that could trigger this error message. (And that hasn't happened for a long time, since PR #1496 and some related work.

**Comment by tsingh2k15** [comment]
> Sorting nodes is a step that determines what order to lint the models. IIRC, in order to lint ephemeral models, it is necessary to first lint at least one other model that depends on the ephemeral model. Although you mentioned there are no ephemeral models in your project, ephemeral models are the only thing I know that could trigger this error message. (And that hasn't happened for a long time, since PR #1496 and some related work.

Thanks for further clarification. In my case, my staging models are using default materialization of `view`.  My intermediate models are `ephemeral` but they are passing the lint. The issue occurs on `staging` models.

**Comment by barrywhart** [comment]
I wonder if "view" behaves similarly to "ephemeral"? In both cases, the query is not materialized as a table in the database. Are the staging models used by any materialized models?

Offhand, I don't have any ideas how to make this work. SQLFluff is limited by how the dbt compiler works. Sorting nodes is a trick we added to work around limitations on ephemeral models, but it only works if there's eventually a table model that queries from the ephemeral model.

Some possible ideas for a workaround:
* Can you define or override a setting for linting (not for actual dbt compiling) to treat these models as tables (not views)?
* Can you add additional **table** models used only for linting that query from these views?

**Comment by tsingh2k15** [comment]
> 

Let me think more on these suggestions.  Sharing my thoughts on your questions below.

Models with `view` materialization exists as `views` in db;  they are not similar to `ephemeral` materialization in this perspective. Models with `ephemeral` materialization do not exist in database, they are imported as CTE into downstream models without building them.  I recently changed the `materialization` for my `intermediate` models from `view` to `ephemeral` to reduce the unnecessary noise in the database as I did not need to query views for my `intermediate` models. I am just using them for code reuse.

My `staging` models(with view materialization) are used by my `intermediate` models(with ephemeral materialization).
My `intermediate` models are used by my `marts` models. My `marts` models are materialized as `tables`.

**Comment by reubster** [comment]
Barry, from my point of view.. I don't get the error in 0.10.1, and I do get the error in 0.11.0 - which is the release that includes the ephemeral fix. 

> Only thing my staging models are using is dbt source, Could it be that the linting tool is looking for a compiled form of source ? I am just relating it to error message - Asked to compile <class 'dbt.contracts.graph.compiled.CompiledModelNode'> node, but it has no compiled form I do not think sources are compiled as such.

Mine too. 

I have now; 
- re-upgraded to 1.1.0 
- changed to materialisation table
- replaced the dbt source with an explicit object 

I get the same error. I get this error for any model. 

Things I note from the trace; 
- the actual linting works, and the stack trace is written afterwards. I guess the exception might have occurred earlier and only been written to the console after the linting output? 
- the actual linting does show errors for the WHOLE file, it's not hitting a problematic line and then exiting. the last linting fail line - line 126 is near the end of the file
- the output shows "file SG__WEB_USERS.sql:  50%|" immediately before the stack trace... 50% of what? 

Stuck figuring out what to try next. Is it viable for me to download the source and run it in a debug mode?

**Comment by alanmcruickshank** [comment]
@reubster - would you be able to paste the _compiled_ version of the `SG__WEB_USERS.sql` file? I think that will help debug this.

**Comment by alanmcruickshank** [comment]
Looking at what's going on in your file, I don't recognise `sv_utils` so I don't know exactly what's going on in them, but unless they contain a `ref` to an ephemeral model, I'm pretty sure if you change `SG__WEB_USERS.sql` to just:

```sql
with SOURCE as (
  select *
  from {{ ref('SG__WEB_USERS__UNION') }}
)

select * from SOURCE
```

I _think_ you'll still get the same error. If you're able to confirm or deny that - it would be very helpful.

**Comment by alanmcruickshank** [comment]
What I think is going on is that SQLFluff is incorrectly sorting the nodes: it's trying to lint `SG__WEB_USERS__UNION` before `SG__WEB_USERS`, and because the former is ephemeral it will fail because it will only trigger compilation once `SG__WEB_USERS` is linted.

@barrywhart - if I'm right I'm wondering whether we can have some kind of _dead letter queue_ where models which fail for this reason are temporarily held until last. We then try and clear the queue - when they'll either pass or if they still fail then we should skip the file.

**Comment by reubster** [comment]
Compiled version: [SG__WEB_USERS.txt](https://github.com/sqlfluff/sqlfluff/files/9095307/SG__WEB_USERS.txt)

**Comment by reubster** [comment]
with model as just 

```
with SOURCE as (
  select *
  from {{ ref('SG__WEB_USERS__UNION') }}
)

select * from SOURCE

```
> sqlfluff lint .\models\base\segment\SG__WEB_USERS.sql
==== sqlfluff ====
sqlfluff:                1.1.0 python:                 3.9.12
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:                 dbt
dbt:                   =0.21.1
rules:                                all
==== readout ====

=== [ path: .\models\base\segment\SG__WEB_USERS.sql ] ===

=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
== [models\base\segment\SG__WEB_USERS.sql] PASS
file SG__WEB_USERS.sql:  50%|█████████████████████████████████████████████████████████████████████████████████                                                                                 | 1/2 [00:11<00:11, 11.46s/it]Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\source\analytics-dbt-models\.venv\Scripts\sqlfluff.exe\__main__.py", line 7, in <module>
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\click\core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\click\core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\click\core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\click\core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\cli\commands.py", line 527, in lint
    result = lnt.lint_paths(
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\linter.py", line 1143, in lint_paths
    self.lint_path(
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\linter.py", line 1095, in lint_path
    for i, linted_file in enumerate(runner.run(fnames, fix), start=1):
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\runner.py", line 101, in run
    for fname, partial in self.iter_partials(fnames, fix=fix):
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\runner.py", line 54, in iter_partials
    for fname, rendered in self.iter_rendered(fnames):
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\runner.py", line 43, in iter_rendered
    yield fname, self.linter.render_file(fname, self.config)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\linter.py", line 816, in render_file
    return self.render_string(raw_file, fname, config, encoding)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\linter.py", line 787, in render_string
    templated_file, templater_violations = self.templater.process(
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff_templater_dbt\templater.py", line 330, in process
    processed_result = self._unsafe_process(fname_absolute_path, in_str, config)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff_templater_dbt\templater.py", line 463, in _unsafe_process
    node = self.dbt_compiler.compile_node(
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\dbt\compilation.py", line 544, in compile_node
    node = self._compile_node(node, manifest, extra_context)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\dbt\compilation.py", line 378, in _compile_node
    compiled_node = _compiled_type_for(node).from_dict(data)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\dbt\compilation.py", line 39, in _compiled_type_for
    raise InternalException(
dbt.exceptions.InternalException: Asked to compile <class 'dbt.contracts.graph.compiled.CompiledModelNode'> node, but it has no compiled form

**Comment by reubster** [comment]
... however if I change the model to just; 

`select 1 `

I get the same error. 

Surely this is a stack thing? I'm running on Windows...?

**Comment by reubster** [comment]
Sorry, also, just tried upgrading to dbt 1.1.1 ... and I get the same error.

**Comment by alanmcruickshank** [comment]
Ok - I think the error message is throwing us off the scent. The issue is not `SG__WEB_USERS` because it's successfully linted that file (see `== [models\base\segment\SG__WEB_USERS.sql] PASS` in the logs).

I'm going to make a branch which has some better logging for this and which will display a more useful output...

**Comment by alanmcruickshank** [comment]
Ok - unfortunately I don't think git supports installing direct from subdirectories of git repositories (you'll need a new version of `sqlfluff-templater-dbt` not the root `sqlfluff` project).

Can you [download the code for that branch](https://github.com/sqlfluff/sqlfluff/archive/refs/heads/ac/dbt_fail_logging.zip) and install from the local version.

You can install the templater locally by running:

```
pip install <path to wherever you extracted the zip file>/plugins/sqlfluff-templater-dbt
```

What you should get instead of getting `dbt.exceptions.InternalException` directly is that dbt will skip the file and carry on - but will also give you a warning with the details of exactly which file (or _files_) are causing the issue. This should at least enable you to use the tool - but also know which files are causing issues.

**Comment by alanmcruickshank** [comment]
The changes I've suggested on that branch above are actually now merged into the main branch (we decided they're actually broadly useful and applicable beyond just debugging). I'll see if we can cut a new release soon which we can then use for testing without needing to install from source.

**Comment by alanmcruickshank** [comment]
Excellent.

If you want an easier setup, the same code is now live as sqlfluff 1.2.0: https://github.com/sqlfluff/sqlfluff/releases/tag/1.2.0

As for the error log. I'm really confused. It appears that `SG__WEB_USERS.sql` gets a `PASS`, `FAIL` *AND* `Skip`...

Can you get the log output for `sqlfluff lint models\base\segment\SG__WEB_USERS.sql -vvvv --logger templater`? As a file is fine - which should give us a little more detail from the templater. From the logs above it appears that the same file is being processed multiple times - which I don't understand so I feel like I need to understand more.

**Comment by alanmcruickshank** [comment]
> Not sure this is helpful either.

I tend to agree. The one thing that it does tell me is that we're definitely processing `models\base\segment\SG__WEB_USERS.sql` more than once. The lines with:

```
DEBUG          Trailing newline count in source dbt model: 1
...
```

...come _after_ the place where the compilation error would be raised. This means that we're _first_ successfully processing the file, and then  for some reason _also_ failing to process the file immediately after.

Most of the handling of which file is linted in which order (and probably the cause of why a file might be linted more than once) is in the `linter` rather than the `templater`. With that in mind, might I ask another logging request: 

```
sqlfluff lint models\base\segment\SG__WEB_USERS.sql -vvvvv --logger linter
```

That will at the very least confirm definitively whether we're attempting to process the same file more than once with different outcomes, or whether we're getting the wrong end of the stick entirely.

**Comment by tsingh2k15** [comment]
Thought to share more findings. I ran the `sqlfluff lint` with  `-vvvvv --logger linter`  option enabled and observed that the issue `only` happens on `staging` models(atleast for my setup). 

Looking at logs of the other models where linter pass, the sequence I observed is:

```
Sorting Nodes...
TEMPLATING RAW
LEXING RAW
PARSING
LINTING
```

For `staging` model, looking at stack trace(below) and found that the error occurs during the first step i.e. `TEMPLATING RAW` while sorting nodes for `staging` model.  

My staging model is really simple, it selects the data from source with template like

`select * from {{ source('example', 'example') }}`

The error originates from https://github.com/dbt-labs/dbt-core/blob/064d89017282cd88fc8d030609d955ce891440a5/core/dbt/compilation.py#L42 where the linter seems to be passing the node as type  `dbt.contracts.graph.compiled.CompiledModelNode` but the expected node type is of `ParsedNode`. These are the expected types in the key of `COMPILED_TYPES`  - https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/core/dbt/contracts/graph/compiled.py#L156

I am not sure why it would only happen on `staging` model. Need to be able to debug the code locally to understand more.
I also verified there is no duplicate linting. Each file is only linted once.

Below is the full trace (replaced some confidential data with `example`).  I will try to install the code locally and debug to find more.


Below are the logs, I could not capture all the logs from terminal;  so only pasting relevant portion.

```

==== readout ====

=== [ path: dbt ] ===

=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
path dbt:   0% 0/22 [00:01<?, ?it/s]                         21:22:05  Unable to do partial parsing because config vars, config profile, or config target have changed [00:00<?, ?it/s]

== [dbt/models/intermediate/int_example_1.sql] LINTING
== [dbt/models/intermediate/int_example_1.sql] PASS
== [dbt/models/intermediate/int_example_2.sql] LINTING
== [dbt/models/intermediate/int_example_2.sql] PASS
== [dbt/models/intermediate/int_example_3.sql] LINTING
== [dbt/models/intermediate/int_example_3.sql] PASS
== [dbt/models/intermediate/int_example_4.sql] LINTING
== [dbt/models/intermediate/int_example_4.sql] PASS
== [dbt/models/intermediate/int_example_5.sql] LINTING
== [dbt/models/intermediate/int_example_5.sql] PASS
== [dbt/models/intermediate/int_example_6.sql] LINTING
== [dbt/models/intermediate/int_example_6.sql] PASS
== [dbt/models/intermediate/int_example_7.sql] LINTING
== [dbt/models/intermediate/int_example_7.sql] PASS
== [dbt/models/intermediate/int_example_8.sql] LINTING
== [dbt/models/intermediate/int_example_8.sql] PASS
== [dbt/models/intermediate/int_example_9.sql] LINTING
== [dbt/models/intermediate/int_example_9.sql] PASS
== [dbt/models/intermediate/int_example_10.sql] LINTING
== [dbt/models/intermediate/int_example_10.sql] PASS
== [dbt/models/marts/example_1.sql] LINTING
== [dbt/models/marts/example_1.sql] PASS
== [dbt/models/marts/example_2.sql] LINTING
== [dbt/models/marts/example_2.sql] PASS



== [dbt/models/staging/example.sql] LINTING
== [dbt/models/staging/example.sql] PASS
=== [ path: dbt/models/staging/example.sql ] ===

=== [dbt templater] Sorting Nodes...
INFO       TEMPLATING RAW [dbt] (dbt/models/staging/example.sql)
path dbt/models/staging/example.sql:  18% 4/22 [00:14<00:38,  2.13s/it]Traceback (most recent call last):
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/cli/commands.py", line 527, in lint
    result = lnt.lint_paths(
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 1143, in lint_paths
    self.lint_path(
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 1095, in lint_path
    for i, linted_file in enumerate(runner.run(fnames, fix), start=1):
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 101, in run
    for fname, partial in self.iter_partials(fnames, fix=fix):
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 54, in iter_partials
    for fname, rendered in self.iter_rendered(fnames):
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 43, in iter_rendered
    yield fname, self.linter.render_file(fname, self.config)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 816, in render_file
    return self.render_string(raw_file, fname, config, encoding)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 787, in render_string
    templated_file, templater_violations = self.templater.process(
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff_templater_dbt/templater.py", line 330, in process
    processed_result = self._unsafe_process(fname_absolute_path, in_str, config)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/sqlfluff_templater_dbt/templater.py", line 463, in _unsafe_process
    node = self.dbt_compiler.compile_node(
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/dbt/compilation.py", line 513, in compile_node
    node = self._compile_node(node, manifest, extra_context)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/dbt/compilation.py", line 370, in _compile_node
    compiled_node = _compiled_type_for(node).from_dict(data)
  File "/Users/example/.cache/pre-commit/repooxli8ycu/py_env-python3.8/lib/python3.8/site-packages/dbt/compilation.py", line 41, in _compiled_type_for
    raise InternalException(
dbt.exceptions.InternalException: Asked to compile <class 'dbt.contracts.graph.compiled.CompiledModelNode'> node, but it has no compiled form

```

**Comment by tsingh2k15** [comment]
Further , I added print statement for https://github.com/sqlfluff/sqlfluff/blob/main/plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py#L384 with `print(results)` to check the contents of `results` and I can confirm that it is indeed picking the node of type `CompiledModelNode` for the failing `staging` model. For passing models, it picks up the correct type i.e. `ParsedModelNode`

It may be a good idea to check the `selector`. Though it looks like this is now related to earlier comments on sorting, if the `staging` model was already compiled and linted, then there may not be any parsed object available? and the only available node type is `CompiledModelNode`? I see that the `results[0]` expects the first entry to be of correct type.

**Comment by alanmcruickshank** [comment]
@tsingh2k15 - if you're happy to add some more `print` statements, you might be able to help a lot in diagnosing where the issue is. 🙏 

The templater defines `.sequence_files()` which dictates which order to yield files in - and you'll see in the final part of that method for the dbt templater there is code to explicitly only yield each file once.

However - in the logs that @tsingh2k15 posted, I can see `=== [dbt templater] Sorting Nodes...` more than once - which could be the cause of emitting the file more than once. For a dbt project that *should* only happen once. If we're caching the project once, but calling `sequence_files()` more than once, that could definitely lead to a bug.

For people experiencing this bug, are you passing multiple commands to `sqlfluff` - incidentally are you running dbt locally or with a grpc server?

**Comment by alanmcruickshank** [comment]
@jtcohen6 - any ideas why we might get `ParsedModelNode` normally, but occasionally get a `CompiledModelNode` return from `Compiler.compile_node()`?

**Comment by reubster** [comment]
@alanmcruickshank 

> For people experiencing this bug, are you passing multiple commands to sqlfluff - incidentally are you running dbt locally or with a grpc server?

linting just the single file. running dbt locally. 

Results using the linter logging; 

[fluff.txt](https://github.com/sqlfluff/sqlfluff/files/9130653/fluff.txt)

**Comment by tsingh2k15** [comment]
> added

@alanmcruickshank It only returns one result. It does not return the right one.  Your suggestion makes sense.

On related note, I updated the pre-commit hook to rev `1.2.0` and I did not see this issue. I will try again to make sure my observation is correct. Are you aware of changes in `1.2.0` that could be related?

**Comment by alanmcruickshank** [comment]
Hold up - it looks like the reason that things are failing is that we're trying to compile a node... _which has already been compiled_. Maybe the solution here is that we just use it _as is_ and don't recompile.

I'm going to see if that might work, and might have a version for you to test shortly...

**Comment by alanmcruickshank** [comment]
@barrywhart I might need your help here.

I think we have three possible options here:
1. Don't recompile the file if `node` is already a `CompiledModelNode`. This feels like the easiest, but I'm not sure how the tracing monkey patch would function properly in that case.
2. Cache the results of `_unsafe_process()`, so we never re-process a file. That dodges the problem here, but potentially leaves us open to issues if we ever get a `CompiledModelNode` on the first attempt.
3. Don't try and fix anything here and instead, focus on why we're trying to process the same file twice.

**Comment by alanmcruickshank** [comment]
Ok - attempting to solve the duplicated node issue there's a PR live which I'd love if both @reubster and @tsingh2k15 would be able to test on their projects: https://github.com/sqlfluff/sqlfluff/pull/3629

You'll need to download the code and install locally.

Based on the logs I'm seeing, I'm relatively confident that it should solve the issues that @reubster is seeing, but I'm less sure about the issues that @tsingh2k15 is seeing.

**Comment by alanmcruickshank** [comment]
If @tsingh2k15 is still seeing issues - I think this might be an issue with the introduction of partial parsing on the dbt side (rather than a change in the dbt templater versions). If they're able to yield a pre-compiled model from their side, we might have to rethink how we do tracing to cope with it.

**Comment by alanmcruickshank** [comment]
> Are you aware of changes in `1.2.0` that could be related?

Looking at git blame: https://github.com/sqlfluff/sqlfluff/blame/main/plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py

No significant changes look like they've been made to the templater in months.

**Comment by barrywhart** [comment]
@reubster: Yes. More specifically, I'm asking if the issue appeared after upgrading dbt. If so, downgrading may fix it. SQLFluff interacts closely with dbt in undocumented, unsupported ways, so dbt updates have been known to break it from time to time. This happened frequently prior to dbt 1.0, and could still happen.

**Comment by jtcohen6** [comment]
I haven't had a chance to dig in very deep here, but just in case it's helpful:

As I understand it, the intended behavior of the [`Compiler.compile_node`](https://github.com/dbt-labs/dbt-core/blob/e03d35a9fc606f5b1fc7e94c0bc04dbe88f558cf/core/dbt/compilation.py#L502-L520) method in `dbt-core` is to accept a parsed node and return a compiled node.

The nomenclature here is a bit confusing:
- The `node` argument is of type [`ManifestNode`](https://github.com/dbt-labs/dbt-core/blob/99bc29258841514d4cf1e911d5f12789927e4a99/core/dbt/contracts/graph/compiled.py#L216-L219), which can be parsed (= not-yet-compiled) or compiled (= already compiled)
- The [`_compiled_type_for`](https://github.com/dbt-labs/dbt-core/blob/e03d35a9fc606f5b1fc7e94c0bc04dbe88f558cf/core/dbt/compilation.py#L39-L44) check here returns the CompiledNode class we're expecting for the ParsedNode class input — and raises an error if it's not a ParsedNode class, i.e. already compiled
- This code can get gnarly for ephemeral models and models that depend on them, for which `compile_node` involves calling a recursive function `_recursively_prepend_ctes`. But that only happens after `_compile_node` is already called, and that's where the `_compiled_type_for` check happens. So I'm not sure how this can happen, except by calling `compile_node` twice on the same manifest resource.

**Comment by reubster** [comment]
Oh. no the issue only appeared after I upgraded sqlfluff. I don't have the issue on 0.10.1 with dbt 0.21.1.. but occurs with every combination of sqlfluff with dbt-core higher than this. 

I guess - the 3rd leg at play here is snowflake?

**Comment by reubster** [comment]
@alanmcruickshank - I downloaded and installed PR3629. I'm getting a stack dump running that. 

> sqlfluff lint models\base\segment\SG__WEB_USERS.sql
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\source\analytics-dbt-models\.venv\Scripts\sqlfluff.exe\__main__.py", line 4, in <module>
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\__init__.py", line 6, in <module>
    from sqlfluff.api import lint, fix, parse, list_rules, list_dialects
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\api\__init__.py", line 4, in <module>
    from sqlfluff.api.simple import lint, fix, parse, APIParsingError
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\api\simple.py", line 4, in <module>
    from sqlfluff.core import (
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\__init__.py", line 9, in <module>
    from sqlfluff.core.linter import Linter
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\__init__.py", line 6, in <module>
    from sqlfluff.core.linter.linter import Linter
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\linter\linter.py", line 23, in <module>
    from sqlfluff.core.rules import get_ruleset
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\rules\__init__.py", line 12, in <module>
    for plugin_rules in get_plugin_manager().hook.get_rules():
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff\core\plugin\host.py", line 12, in get_plugin_manager
    pm.load_setuptools_entrypoints(project_name)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\pluggy\_manager.py", line 287, in load_setuptools_entrypoints
    plugin = ep.load()
  File "C:\Program Files\Python310\lib\importlib\metadata\__init__.py", line 171, in load
    module = import_module(match.group('module'))
  File "C:\Program Files\Python310\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff_templater_dbt\__init__.py", line 3, in <module>
    from sqlfluff_templater_dbt.templater import DbtTemplater
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\sqlfluff_templater_dbt\templater.py", line 12, in <module>
    from dbt.version import get_installed_version
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\dbt\version.py", line 10, in <module>
    import dbt.exceptions
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\dbt\exceptions.py", line 5, in <module>
    from dbt.logger import GLOBAL_LOGGER as logger
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\dbt\logger.py", line 16, in <module>
    from dbt.dataclass_schema import dbtClassMixin
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\dbt\dataclass_schema.py", line 13, in <module>
    from mashumaro import DataClassDictMixin
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\__init__.py", line 4, in <module>
    from mashumaro.serializer.json import DataClassJSONMixin
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\serializer\json.py", line 28, in <module>
    class DataClassJSONMixin(DataClassDictMixin):
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\serializer\base\dict.py", line 23, in __init_subclass__
    raise exc
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\serializer\base\dict.py", line 19, in __init_subclass__
    builder.add_to_dict()
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\serializer\base\metaprogramming.py", line 379, in add_to_dict
    pre_serialize = self.get_declared_hook(__PRE_SERIALIZE__)
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\serializer\base\metaprogramming.py", line 206, in get_declared_hook
    if not is_dataclass_dict_mixin(cls):
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\meta\helpers.py", line 88, in is_dataclass_dict_mixin
    return type_name(t) == DataClassDictMixinPath
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\meta\helpers.py", line 21, in type_name
    if is_generic(t):
  File "C:\source\analytics-dbt-models\.venv\lib\site-packages\mashumaro\meta\helpers.py", line 47, in is_generic
    raise NotImplementedError
NotImplementedError

**Comment by alanmcruickshank** [comment]
> I guess - the 3rd leg at play here is snowflake?

I think it's very unlikely that this is a snowflake issue.

**Comment by alanmcruickshank** [comment]
> I downloaded and installed PR3629. I'm getting a stack dump running that.

@reubster - that's a very surprising issue. It looks like it's an issue with importing `mashumaro` before any work is done at all in dbt or in sqlfluff. Given the changes made in that branch, where there are no changes to any of the imports, it seems strange that we would have triggered an import error.

One thing I do know is that support for `dbt-snowflake` in python 310 isn't live yet (see supported versions on [pypi](https://pypi.org/project/dbt-snowflake/)), and part of that is due to dependencies, so while it's a bit of a shot in the dark, would you be able to try the same thing but in python 3.7, 3.8 or 3.9? I can see that your testing env is a python 3.10 env. I know 3.10 is coming for that package but it's still in pre-release.

**Comment by reubster** [comment]
@alanmcruickshank 

I completely reset my virtual environment and reran it (PR3629). this works!  I've tried various models and the errors are gone. 

stack & output 

```
==== sqlfluff ====
sqlfluff:                1.2.1 python:                 3.9.12
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:                 dbt
dbt:                   =0.21.1
rules:                                all
==== readout ====

=== [ path: .\models\base\segment\SG__WEB_USERS.sql ] ===

=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
== [models\base\segment\SG__WEB_USERS.sql] FAIL
L:   4 | P:   3 | L036 | Select targets should be on a new line unless there is
                       | only one select target.
L:   8 | P:   2 | L019 | Found trailing comma. Expected only leading.
L:  12 | P:  25 | L019 | Found trailing comma. Expected only leading.
L:  13 | P:  10 | L019 | Found trailing comma. Expected only leading.
L:  15 | P:  61 | L019 | Found trailing comma. Expected only leading.
L:  17 | P:  61 | L019 | Found trailing comma. Expected only leading.
L:  19 | P:  18 | L019 | Found trailing comma. Expected only leading.
L:  20 | P:  14 | L019 | Found trailing comma. Expected only leading.
L:  21 | P:  13 | L019 | Found trailing comma. Expected only leading.
L:  22 | P:  18 | L019 | Found trailing comma. Expected only leading.
L:  23 | P:  77 | L019 | Found trailing comma. Expected only leading.
L:  24 | P:  77 | L019 | Found trailing comma. Expected only leading.
L:  25 | P:  24 | L019 | Found trailing comma. Expected only leading.
L:  26 | P:  15 | L019 | Found trailing comma. Expected only leading.
L:  27 | P:  15 | L019 | Found trailing comma. Expected only leading.
L:  28 | P:  13 | L019 | Found trailing comma. Expected only leading.
L:  29 | P:  12 | L019 | Found trailing comma. Expected only leading.
L:  30 | P:  14 | L019 | Found trailing comma. Expected only leading.
L:  31 | P:  14 | L019 | Found trailing comma. Expected only leading.
L:  33 | P:  21 | L019 | Found trailing comma. Expected only leading.
L:  35 | P:  18 | L019 | Found trailing comma. Expected only leading.
L:  36 | P:  42 | L019 | Found trailing comma. Expected only leading.
L:  37 | P:  33 | L019 | Found trailing comma. Expected only leading.
L:  38 | P:  70 | L019 | Found trailing comma. Expected only leading.
L:  40 | P:  43 | L019 | Found trailing comma. Expected only leading.
L:  43 | P:  60 | L019 | Found trailing comma. Expected only leading.
L:  44 | P:  13 | L019 | Found trailing comma. Expected only leading.
L:  46 | P:  28 | L019 | Found trailing comma. Expected only leading.
L:  48 | P:  37 | L019 | Found trailing comma. Expected only leading.
L:  49 | P:  29 | L019 | Found trailing comma. Expected only leading.
L:  50 | P:  35 | L019 | Found trailing comma. Expected only leading.
L:  51 | P:  42 | L019 | Found trailing comma. Expected only leading.
L:  52 | P:  42 | L019 | Found trailing comma. Expected only leading.
L:  53 | P:  42 | L019 | Found trailing comma. Expected only leading.
L:  54 | P:  13 | L019 | Found trailing comma. Expected only leading.
L:  55 | P:  37 | L019 | Found trailing comma. Expected only leading.
L:  57 | P:  18 | L014 | Unquoted identifiers must be consistently upper case.
L:  61 | P:  11 | L019 | Found trailing comma. Expected only leading.
L:  62 | P:  13 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  63 | P:  12 | L014 | Unquoted identifiers must be consistently upper case.
L:  63 | P:  18 | L019 | Found trailing comma. Expected only leading.
L:  64 | P:  12 | L019 | Found trailing comma. Expected only leading.
L:  65 | P:  41 | L019 | Found trailing comma. Expected only leading.
L:  66 | P:  27 | L019 | Found trailing comma. Expected only leading.
L:  67 | P:  14 | L019 | Found trailing comma. Expected only leading.
L:  68 | P:  20 | L019 | Found trailing comma. Expected only leading.
L:  69 | P:  15 | L019 | Found trailing comma. Expected only leading.
L:  70 | P:  12 | L019 | Found trailing comma. Expected only leading.
L:  71 | P:  16 | L019 | Found trailing comma. Expected only leading.
L:  72 | P:  22 | L019 | Found trailing comma. Expected only leading.
L:  73 | P:  17 | L019 | Found trailing comma. Expected only leading.
L:  74 | P:  32 | L019 | Found trailing comma. Expected only leading.
L:  75 | P:  31 | L019 | Found trailing comma. Expected only leading.
L:  76 | P:  23 | L019 | Found trailing comma. Expected only leading.
L:  77 | P:  26 | L019 | Found trailing comma. Expected only leading.
L:  79 | P:  13 | L019 | Found trailing comma. Expected only leading.
L:  80 | P:  18 | L019 | Found trailing comma. Expected only leading.
L:  81 | P:  26 | L019 | Found trailing comma. Expected only leading.
L:  82 | P:  19 | L019 | Found trailing comma. Expected only leading.
L:  83 | P:  20 | L019 | Found trailing comma. Expected only leading.
L:  84 | P:  19 | L019 | Found trailing comma. Expected only leading.
L:  85 | P:  28 | L019 | Found trailing comma. Expected only leading.
L:  86 | P:  15 | L019 | Found trailing comma. Expected only leading.
L:  87 | P:  25 | L019 | Found trailing comma. Expected only leading.
L:  88 | P:  26 | L019 | Found trailing comma. Expected only leading.
L:  89 | P:  16 | L019 | Found trailing comma. Expected only leading.
L:  90 | P:  27 | L019 | Found trailing comma. Expected only leading.
L:  91 | P:  30 | L019 | Found trailing comma. Expected only leading.
L:  93 | P:  10 | L019 | Found trailing comma. Expected only leading.
L:  94 | P:  24 | L019 | Found trailing comma. Expected only leading.
L:  95 | P:  31 | L019 | Found trailing comma. Expected only leading.
L:  96 | P:  28 | L019 | Found trailing comma. Expected only leading.
L:  97 | P:  23 | L019 | Found trailing comma. Expected only leading.
L:  98 | P:  37 | L019 | Found trailing comma. Expected only leading.
L:  99 | P:  37 | L019 | Found trailing comma. Expected only leading.
L: 100 | P:  34 | L019 | Found trailing comma. Expected only leading.
L: 101 | P:  29 | L019 | Found trailing comma. Expected only leading.
L: 103 | P:  31 | L019 | Found trailing comma. Expected only leading.
L: 104 | P:  37 | L019 | Found trailing comma. Expected only leading.
L: 105 | P:  18 | L019 | Found trailing comma. Expected only leading.
L: 106 | P:  14 | L019 | Found trailing comma. Expected only leading.
L: 107 | P:  25 | L019 | Found trailing comma. Expected only leading.
L: 108 | P:  10 | L019 | Found trailing comma. Expected only leading.
L: 109 | P:  17 | L019 | Found trailing comma. Expected only leading.
L: 111 | P:  18 | L019 | Found trailing comma. Expected only leading.
L: 112 | P:  25 | L019 | Found trailing comma. Expected only leading.
L: 113 | P:  27 | L019 | Found trailing comma. Expected only leading.
L: 114 | P:  23 | L019 | Found trailing comma. Expected only leading.
L: 115 | P:  22 | L019 | Found trailing comma. Expected only leading.
L: 116 | P:  25 | L019 | Found trailing comma. Expected only leading.
L: 117 | P:  26 | L019 | Found trailing comma. Expected only leading.
L: 122 | P:   2 | L019 | Found trailing comma. Expected only leading.
L: 125 | P:   9 | L001 | Unnecessary trailing whitespace.
L: 126 | P:   6 | L019 | Found trailing comma. Expected only leading.
L: 138 | P:   1 | L036 | Select targets should be on a new line unless there is
                       | only one select target.
==== summary ====
violations:       95 status:         FAIL
```

**Comment by alanmcruickshank** [comment]
Awesome. Thanks @reubster!

@tsingh2k15 - would you be able to test this with your setup too? From what I saw of your logs I wonder if your scenario might be more complicated.

@barrywhart - can you think of any reasons that it would be a bad idea to merge #3629 ? It feels like quite aggressive deduplication, and might be dodging the core issue here but at least from the feedback above it seems to be an effective fix.

**Comment by barrywhart** [comment]
Seems fine to merge the PR. I'm curious, is there logging in place to tell when it sees a duplicate? Thinking that might be useful sometimes for further debugging, perhaps?

**Comment by alanmcruickshank** [comment]
> is there logging in place to tell when it sees a duplicate?

No, but that's a good idea. I'll do that and then merge the PR.

**Comment by alanmcruickshank** [comment]
Merged. @reubster your issues will be fixed in the next release 👍.

@tsingh2k15 I'll leave this issue open until you've checked whether your project is also working with these fixes. You can either download from the now merged branch or you can download from the current `main`.

**Comment by tsingh2k15** [comment]
@alanmcruickshank Sorry for the late reply. I was able to unblock myself by upgrading to `1.2.0` and then did not get chance to follow up on this thread. I am not sure why the issue did not occur in `1.2.0`, it could be very likely some staleness problem with my local setup which may have resolved after upgrading to `1.2.0`. I will also plan to test the fixes but the changes look good to me, feel free to close this issue.
